File: app/src/main/java/com/shagalalab/qqkeyboard/settings/vibration/VibrationLevelPreferenceDialogFragment.java
package com.shagalalab.qqkeyboard.settings.vibration;

import android.content.Context;
import android.os.Bundle;
import android.os.Vibrator;
import android.view.View;
import android.widget.SeekBar;
import android.widget.TextView;

import com.shagalalab.qqkeyboard.R;

import androidx.preference.DialogPreference;
import androidx.preference.PreferenceDialogFragmentCompat;

public class VibrationLevelPreferenceDialogFragment extends PreferenceDialogFragmentCompat
    implements SeekBar.OnSeekBarChangeListener {

    private TextView caption;
    private SeekBar seekBar;
    private Vibrator vibrator;

    public static VibrationLevelPreferenceDialogFragment newInstance(String key) {
        final VibrationLevelPreferenceDialogFragment fragment = new VibrationLevelPreferenceDialogFragment();
        final Bundle b = new Bundle(1);
        b.putString(ARG_KEY, key);
        fragment.setArguments(b);
        return fragment;
    }

    @Override
    protected void onBindDialogView(View view) {
        caption = view.findViewById(R.id.vibration_level_preference_caption);
        seekBar = view.findViewById(R.id.vibration_level_preference_seekbar);
        seekBar.setOnSeekBarChangeListener(this);

        Integer vibrationLevel = null;
        DialogPreference preference = getPreference();
        if (preference instanceof VibrationLevelPreference) {
            vibrationLevel = ((VibrationLevelPreference) preference).getVibrationLevel();
        }

        if (vibrationLevel != null) {
            seekBar.setProgress(vibrationLevel);
        }
    }

    @Override
    public void onDialogClosed(boolean positiveResult) {
        if (positiveResult) {
            int level = seekBar.getProgress();
            setCaption(level);

            DialogPreference preference = getPreference();
            if (preference instanceof VibrationLevelPreference) {
                VibrationLevelPreference vibrationPreference = (VibrationLevelPreference) preference;
                if (vibrationPreference.callChangeListener(level)) {
                    vibrationPreference.setVibrationLevel(level);
                    vibrationPreference.setSummary(String.format(requireContext().getString(R.string.text_with_ms), level));
                }
            }
        }
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        setCaption(progress);
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {

    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
        getVibrator().vibrate(seekBar.getProgress());
    }

    private Vibrator getVibrator() {
        if (vibrator == null) {
            vibrator = (Vibrator) getContext().getSystemService(Context.VIBRATOR_SERVICE);
        }

        return vibrator;
    }

    private void setCaption(int progress) {
        caption.setText(String.format(requireContext().getString(R.string.text_with_ms), progress));
    }
}

File: app/src/main/java/com/shagalalab/qqkeyboard/settings/vibration/VibrationLevelPreferenceDialogFragment.kt
package com.shagalalab.qqkeyboard.settings.vibration

import android.content.Context
import android.os.Bundle
import android.os.Vibrator
import android.view.View
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import androidx.preference.PreferenceDialogFragmentCompat
import com.shagalalab.qqkeyboard.R


class VibrationLevelPreferenceDialogFragment : PreferenceDialogFragmentCompat(),
    OnSeekBarChangeListener {
    private var caption: TextView? = null
    private lateinit var seekBar: SeekBar
    private var vibrator: Vibrator? = null
        get() {
            if (field == null) {
                field = requireContext().getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
            }

            return field
        }

    override fun onBindDialogView(view: View) {
        caption = view.findViewById<TextView>(R.id.vibration_level_preference_caption)
        seekBar = view.findViewById<SeekBar>(R.id.vibration_level_preference_seekbar)
        seekBar.setOnSeekBarChangeListener(this)

        var vibrationLevel: Int? = null
        val preference = preference
        if (preference is VibrationLevelPreference) {
            vibrationLevel = preference.vibrationLevel
        }

        if (vibrationLevel != null) {
            seekBar.setProgress(vibrationLevel)
        }
    }

    override fun onDialogClosed(positiveResult: Boolean) {
        if (positiveResult) {
            val level = seekBar.progress
            setCaption(level)

            val preference = preference
            if (preference is VibrationLevelPreference) {
                if (preference.callChangeListener(level)) {
                    preference.vibrationLevel = level
                    preference.summary =
                        String.format(requireContext().getString(R.string.text_with_ms), level)
                }
            }
        }
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
        setCaption(progress)
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {
    }

    override fun onStopTrackingTouch(seekBar: SeekBar) {
        vibrator!!.vibrate(seekBar.progress.toLong())
    }

    private fun setCaption(progress: Int) {
        caption!!.text =
            String.format(requireContext().getString(R.string.text_with_ms), progress)
    }

    companion object {
        fun newInstance(key: String?): VibrationLevelPreferenceDialogFragment {
            val fragment = VibrationLevelPreferenceDialogFragment()
            val b = Bundle(1)
            b.putString(ARG_KEY, key)
            fragment.arguments = b
            return fragment
        }
    }
}
